TypeCobol/Compiler/Sql/Scanner/SqlScanner.cs

```diff
@@ -50,46 +80,248 @@ public override Token GetTokenStartingFrom(int startIndex)
+
+                    CheckDecimalFloatingPointLiteral();
+                }
+            }
+
+            void CheckDecimalFloatingPointLiteral()
+            {
+                Debug.Assert(token.LiteralValue is DecimalFloatingPointLiteralTokenValue);
+                var literalValue = (DecimalFloatingPointLiteralTokenValue)token.LiteralValue;
+
+                //Check exponent length
+                var exponent = literalValue.NumericConstant.Exponent.Number;
+                var mantissa = literalValue.NumericConstant.Mantissa.Number;
+                int mantissaLength =
+                    mantissa.ToString().Length;
+                int exponentLength =
+                    exponent.ToString().Length -
+                    (exponent.Sign == -1 ? 1 : 0); //Check digits only, so remove '-' at beginning of negative numbers
+                if (exponentLength > 4 || mantissaLength > 34)
                 {
-                    case TokenType.SQL_COMMIT:
-                    case TokenType.SQL_SELECT:
-                    case TokenType.SQL_ALL:
-                    case TokenType.SQL_DISTINCT:
-                    case TokenType.UserDefinedWord:
-                    case TokenType.SQL_ROLLBACK:
-                    case TokenType.SQL_TO:
-                    case TokenType.SQL_SAVEPOINT:
-                    case TokenType.SQL_FROM:
-                    case TokenType.SQL_AS:
-                    case TokenType.SQL_TABLE:
-                    case TokenType.SQL_DELETE:
-                    case TokenType.SQL_IMMEDIATE:
-                    case TokenType.SQL_DROP:
-                    case TokenType.SQL_RESTRICT:
-                    case TokenType.SQL_WHEN:
-                    case TokenType.SQL_TRUNCATE:
-
-                        return new Token(tokenType, startIndex, currentIndex - 1, tokensLine);
-                    default:
-                        //Unrecognized keyword (for now) return as ExecStatementText
-                        return new Token(TokenType.ExecStatementText, startIndex, currentIndex - 1, tokensLine);
+                    tokensLine.AddDiagnostic(MessageCode.InvalidExponentInFloatingPointLiteral, token);
                 }
             }
-            //Consume all sql-keyword incompatible chars
-            for (; currentIndex <= lastIndex && !IsSqlKeywordPart(line[currentIndex]); currentIndex++){}
-            return new Token(TokenType.ExecStatementText, startIndex, currentIndex - 1, tokensLine);
+
+        }
+
+        public bool TryScanDecimalFloatingPointSpecialValue(int startIndex, bool isNegative,
+            out DecimalFloatingPointLiteralTokenValue decimalFloatingPointLiteralTokenValue)
+
+        {
+            decimalFloatingPointLiteralTokenValue =
+                new DecimalFloatingPointLiteralTokenValue(LiteralTokenValueType.DecimalFloatingPoint, null);
+
+            for (; currentIndex <= lastIndex && IsSqlKeywordPart(line[currentIndex]); currentIndex++)
+            {
+            }
```

Comment:
No. You have to try to find a word but not actually consume the chars. So this is more something like

```csharp
for (int index = startIndex; index <= lastIndex && char.IsLetter(line[index]); index++)
{
}
```

Then build the substring and compare to the values in the dictionary.

Also the boolean `isNegative` is ambuguous. In fact this is more `hasExplicitSign` that indicate that we have to start one char after start index.